### ISSUE TITLE
TComponent raiseEvent Subproperties of handlers calling on event object rather than the handler object.

### ISSUE THREAD

**Issue by belisoful** [pradosoft/prado#733]
https://github.com/pradosoft/prado/blob/690d6a9ac9834752c3a3df65e05816a2c7238624/framework/TComponent.php#L1161

There seems to be a problem with this line of code.  Shouldn't this 

TComponent: L1161			$object = $this->getSubProperty(substr($method, 0, $pos));

be this:

TComponent: L1161			$object = $object->getSubProperty(substr($method, 0, $pos));

?

I checked the history and this seems to be an issue going way back.

This needs a unit test!!!

**Comment by ctrlaltca** [comment]
Nice, looks like you catched a 16-years old bug introduced in 24f744404ab5d148a0bfda03296baa59f7e10725.
In order to trigger it you need to bind and event to a callable with an actual object (not a string) and a method that includes a subproperty of the object, eg 
```
$component->OnClick[]=[$object,'Parent.buttonClicked'];
``` 
This was probably never caught before because in templates there's no easy way to pass object instances to events.
Looks like Yii is not affected by this since they just removed the subproperty handling: https://github.com/yiisoft/yii/blob/eccb713b7dbd07cf2968212bfdf6b469ee818c2e/framework/base/CComponent.php#L561

**Comment by belisoful** [comment]
hmm.   I am looking to write a unit test for this, but ohpunit/composer is failing:

`PHP Warning:  require_once(/Applications/Ampps/prado/tests/test_tools/../../vendor/autoload.php): failed to open stream: No such file or directory in /Applications/Ampps/prado/tests/test_tools/phpunit_bootstrap.php on line 15`

It seems that there is a specific directory that the unit tests are supposed to be run from within.  The PHPUnit Bootstrap seems to be pointing in the wrong place(s).  Is this a bug or a feature?  Any suggestions?

is the phpunit test supposed to work out of the box?  If it is, this is a huge fail for us.

**Comment by ctrlaltca** [comment]
I just tried to run tests and they seems to work (also, they work on CI).
Does the `/Applications/Ampps/prado/vendor/autoload.php` file exists on your mac? if it doesn't you need to run `composer install` from within `/Applications/Ampps/prado`

**Comment by belisoful** [comment]
Ok. so this is where some documentation might be very useful.  How to run unit tests.

My process.  I downloaded Prado via

`git clone https://github.com/pradosoft/prado.git`

make sure phpunit 8 is downloaded and installed. then call:

`composer unittest`

It sounds like you are suggesting that unit testing is for *installed* applications, yes?  maybe this should be documented.

I have an install to test it out upon.  I have a softlink from the internal Prado directory/files to my working Prado directory.  that could be messing up things.

**Comment by belisoful** [comment]
I put put my Prado directories without any links or funkiness.  It seems the directories aren't quite right.

`PHP Warning:  require_once(/Applications/Ampps/www/app/protected/vendor/pradosoft/prado/tests/test_tools/../../vendor/autoload.php): failed to open stream: No such file or directory in /Applications/Ampps/www/app/protected/vendor/pradosoft/prado/tests/test_tools/phpunit_bootstrap.php on line 15`

**Comment by belisoful** [comment]
Selenium also seems to have branched into multiple products.  It's not clear what should be installed for selenium.

After correcting the address, I'm getting phpunit selenium errors from not having selenium installed. FYI.  I don't know what product to install for Selenium.   is there any clarity?

The changes in selenium should be documented as well.

**Comment by ctrlaltca** [comment]
Looking at your paths, these are the steps needed to run tests:
```
cd /Applications/Ampps/
git clone https://github.com/pradosoft/prado.git`
cd prado
composer install
composer unittest
```
The "composer install" step will take care of installing all the needed development tools, including phpunit.

Regarding Selenium, you need the "WebDriver" from this page: https://www.selenium.dev/downloads/ , looks like they call it "Grid" now.
Direct link: https://selenium-release.storage.googleapis.com/3.141/selenium-server-standalone-3.141.59.jar

It's a java application, so you'll need to install a JRE/JDK, the run it from a terminal:
```
java -jar selenium-server-standalone-3.141.59.jar
```

Depending on the browser you want to use to do functional testing, you may need an additional driver: https://www.selenium.dev/documentation/en/webdriver/driver_requirements/

**Comment by belisoful** [comment]
That's perfect.  Can we put that in the READ ME under testing?

**Comment by belisoful** [comment]
Also.  Thank you.  The testing environment is working.  TComponentTest is actually wonky.  The object raising the event is the object in the handler in the test environment.  lol.  that's why it didn't find the bug.  I had a hand in that when I added behaviors, dynamic events, etc to TComponent and didn't re-write the tests properly.

I'll have a check-in on my branch soon.  It'll fix this bug and the WeakReferences Addition, plus a few tweaks to the tests.

**Comment by ctrlaltca** [comment]
> That's perfect. Can we put that in the READ ME under testing?

Done

**Comment by ctrlaltca** [comment]
Fixed in #747
